### ISSUE TITLE
A2C "stuck" on a single action solution

### ISSUE THREAD

**Issue by dklaiman** [tensorforce/tensorforce#781]
Hi,

I am using an A2C agent where it can choose from a discrete set of actions 0, 1, 2, ...n with my own environment.
It tries out in the beginning of training different actions but converges to a single action policy, i.e. regardless of the state it choose always the same action, which is *not* the optimal policy.
(I am training a toy example where the optimal policy is known)

Any ideas why it happens / what should I do?

**Comment by AlexKuhnle** [comment]
Can you post the agent config? Tensorforce's A2C can be very sensitive to some hyperparameters, and you likely need to experiment with them.

**Comment by dklaiman** [comment]
Sure. Thank you for replying so quickly.
I've tried different combinations. Here is an example:

Agent.create(
            agent='a2c',
            environment=sim,
            batch_size = 100,
            horizon = 12, # number of steps per episode
            exploration = 1e-4, # worked with values between 1e-3 - 1e-5, tried also linearly decaying similar to learning rate
            entropy_regularization = 5e-4, # same as exploration
            discount = 0.95,
            learning_rate=dict(type='linear',
                           unit='episodes',
                           num_steps=n_episodes,
                           initial_value=1e-3,
                           final_value=1e-4) # I am usually using decaying value starting with 1e-3
            l2_regularization = 0.02,
            summarizer=dict(directory='./summaries/', max_summaries=1000),
            saver=dict(directory='model-checkpoint', frequency=10, max_checkpoints=5),
            config=dict(seed=42)
        )

**Comment by AlexKuhnle** [comment]
Two things stand out to me:
- First, you probably don't need `exploration` since actions are *sampled* anyway for A2C, and `1e-4` would be extremely rarely anyway. But this shouldn't cause problems, just not necessary.
- Second, your comment for `horizon` suggests that there may be a misunderstanding. `horizon` is not the episode length (which is handled implicitly via the `environment` argument, if your environment implements `max_episode_timesteps`), but it's the return horizon. Do you know how DQN and particularly n-step DQN works (TD, Bellman equation, but unrolled for multiple steps)? `horizon` is basically the same as "n" in n-step DQN.

**Comment by dklaiman** [comment]
I would like the agent to calculate the return over the entire max_episode_timesteps and to maximize this value and this is why I set to be horizon= max_episode_timesteps

**Comment by AlexKuhnle** [comment]
In that case it probably makes sense to switch to the PPO agent, since the horizon < episode-length is one of the main "differences" to the policy-gradient agents (VPG, PPO, TRPO, of which PPO is the recommended choice).

**Comment by dklaiman** [comment]
thank you! I'll try it out

**Comment by AlexKuhnle** [comment]
Assume this is resolved.

**Comment by ibrahimth** [comment]
@dklaiman I have the same problem. Does PPO solve the problem for you?
